sycl/source/detail/config.cpp

```diff
@@ -0,0 +1,117 @@
+//==---------------- config.cpp ---------------------------------*- C++-*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl/detail/os_util.hpp>
+#include <detail/config.hpp>
+
+#include <cstring>
+#include <fstream>
+#include <iostream>
+#include <limits>
+
+#define STRINGIFY_LINE_HELP(s) #s
+#define STRINGIFY_LINE(s) STRINGIFY_LINE_HELP(s)
+
+namespace cl {
+namespace sycl {
+namespace detail {
+
+#ifndef SYCL_CONFIG_FILE_NAME
+#define SYCL_CONFIG_FILE_NAME "sycl.conf"
+#endif // SYCL_CONFIG_FILE_NAME
+
+#define CONFIG(Name, MaxSize, CompileTimeDef)                                  \
+  const char *SYCLConfigBase<Name>::MValueFromFile = nullptr;                  \
+  char SYCLConfigBase<Name>::MStorage[MaxSize];                                \
+  const char *const SYCLConfigBase<Name>::MCompileTimeDef =                    \
+      getStrOrNullptr(STRINGIFY_LINE(CompileTimeDef));                         \
+  const char *const SYCLConfigBase<Name>::MConfigName = STRINGIFY_LINE(Name);
+#include "detail/config.def"
+#undef CONFIG
+
+#define MAX_CONFIG_NAME 256
+
+static void initValue(const char *Key, const char *Value) {
+#define CONFIG(Name, MaxSize, CompileTimeDef)                                  \
+  if (0 == strncmp(Key, SYCLConfigBase<Name>::MConfigName, MAX_CONFIG_NAME)) { \
+    strncpy(SYCLConfigBase<Name>::MStorage, Value, MaxSize);                   \
```

Comment:
And what happens when you reach the limit ? A non-`\0` ending string... :-(

Reply:
You are right. Will fix by guaranteeing \0 at the and of the Value.

Reply:
I would prefer just using safer C++ instead of lower-level C at the first place...